### PR TITLE
comm - started duckdbvectorestore

### PR DESCRIPTION
**Pull request by 3rd-Son**
DuckDb is a traditional database but it supports vector storage using the vss extension which is an experimental extension for DuckDB that adds indexing support to accelerate vector similarity search queries using DuckDB's new fixed-size ARRAY type.

I have implemented this feature request but two test cases are still failing
- test_retrieve
- test_update_document

### Below is the error log

```
======================================================================================== short test summary info =========================================================================================
FAILED pkgs/community/tests/unit/vector_stores/DuckDBVectoreStore_test.py::test_update_document - RuntimeError: Failed to update document test_id: Constraint Error: Duplicate key "id: test_id" violates primary key constraint. If this is an unexpected constraint violation please double check wit...
FAILED pkgs/community/tests/unit/vector_stores/DuckDBVectoreStore_test.py::test_retrieve - RuntimeError: Failed to retrieve similar documents: Binder Error: No function matches the given name and argument types 'array_cosine_distance(FLOAT[100], DOUBLE[])'. You might need to add explicit...
================================================================================= 2 failed, 6 passed, 1 warning in 9.02s =================================================================================
```

To test locally just install `pip install duckdb` and then run the pytest locally

### CODE REVIEW

File: pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+from swarmauri.vector_stores.base.VectorStoreRetrieveMixin import (
+    VectorStoreRetrieveMixin,
+)
+from swarmauri.vector_stores.base.VectorStoreSaveLoadMixin import (
+    VectorStoreSaveLoadMixin,
+)
+
+
+class DuckDBVectorStore(
+    VectorStoreSaveLoadMixin, VectorStoreRetrieveMixin, VectorStoreBase, BaseModel
```
Comment (cobycloud) [inline pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py]: We import BaseModel from ComponentBase

File: pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    client: Optional[duckdb.DuckDBPyConnection] = Field(default=None, exclude=True)
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self, db_path: str, vector_size: int, metric: str = "cosine", **kwargs
+    ):
+        super().__init__(
+            db_path=db_path, vector_size=vector_size, metric=metric, **kwargs
```
Comment (cobycloud) [inline pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py]: If its defined at class then we dont need to define here

**Review by cobycloud — CHANGES_REQUESTED** [review]

File: pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py (hunk 1 above)
Comment (3rdsonx) [reply pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py]: noted
Will do the necessary changes

File: pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py (hunk 2 above)
Comment (3rdsonx) [reply pkgs/community/swarmauri_community/vector_stores/DuckDBVectoreStore.py]: That's true
Thanks for that.
Changes already done

**Comment by cobycloud** [comment]
merge pending small changes

**Review by RafaelJohn9 — CHANGES_REQUESTED** [review]
see issue https://github.com/run-llama/llama_index/issues/15212 to implement DuckDB vector store

**Comment by 3rd-Son** [comment]
> see issue [run-llama/llama_index#15212](https://github.com/run-llama/llama_index/issues/15212) to implement DuckDB vector store

In my last commit, I used the llama_index approach which makes it easier and retains the project structure though the `test_top_k` still fails

**Review by cobycloud — APPROVED** [review]

**Pull request closed without merging**